---
title: "rtplot"
author: "Ruining Dong"
date: "30/10/2019"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(StructuralVariantAnnotation)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
genes <- TxDb.Hsapiens.UCSC.hg19.knownGene
cmhp10.gr <- breakpointRanges(readVcf("~/Documents/CombiMet/CMHP10.sv.vcf"))
gr <- rtDetec(cmhp10.gr, genes, maxgap = 20, minscore = 0.5)
gr$rt
```

```{r ref exons}
txs.name <- unique(unlist(gr$rt$txs))
txs.plot <- transcripts(genes, filter=list(tx_name=txs.name))
exons.plot <- exons(genes, filter=list(tx_name=txs.name), columns=c("exon_id", "tx_name"))
exons.plot$tx_name <- exons.plot$tx_name[sapply(exons.plot$tx_name, function(x){x %in% txs.name})]
tx_name <- c(sapply(exons.plot[sapply(exons.plot$tx_name, length) == 2]$tx_name,function(x){c(x[1])}),
                        sapply(exons.plot[sapply(exons.plot$tx_name, length) == 2]$tx_name,function(x){c(x[2])}),
                        unlist(exons.plot[sapply(exons.plot$tx_name, length) == 1]$tx_name))
exons.plot <- c(exons.plot[sapply(exons.plot$tx_name, length) == 2], 
                exons.plot[sapply(exons.plot$tx_name, length) == 2],
                exons.plot[sapply(exons.plot$tx_name, length) == 1])
exons.plot$tx_name <- tx_name
as_tibble(exons.plot)
```
```{r intronic dels}
intronic.dels <- c(gr$rt[sapply(gr$rt$txs, length) == 1], 
                   gr$rt[sapply(gr$rt$txs, length) == 2], 
                   gr$rt[sapply(gr$rt$txs, length) == 2])
intronic.dels$tx_name <- c(unlist(gr$rt[sapply(gr$rt$txs, length) == 1]$txs),
                           sapply(gr$rt[sapply(gr$rt$txs, length) == 2]$txs, function(x){x[1]}),
                           sapply(gr$rt[sapply(gr$rt$txs, length) == 2]$txs, function(x){x[2]}))
as_tibble(intronic.dels)
intronic.dels.bedpe <- breakpointgr2bedpe(gr$rt) %>% left_join(as_tibble(intronic.dels)[,c("sourceId", "tx_name")], by = c("name" = "sourceId")) 
```


```{r}
ggplot() +
  geom_segment(data=as_tibble(txs.plot), aes(x=start, xend=end, y=1.5, yend=1.5), colour="black") +
  geom_rect(data=as_tibble(exons.plot), mapping=aes(xmin=start, xmax = end, ymin=1, ymax = 2, fill=seqnames), colour="black") +
  # geom_point(data=g, aes(x=start, y=3, colour=factor(call))) +
  geom_curve(data=as_tibble(intronic.dels.bedpe), aes(x=start1, y=2, xend=start2, yend=2), curvature = -0.5, size=1) +
  facet_wrap(~tx_name, scales = "free_x") +
  theme_bw() + scale_y_continuous(limits = c(0, 4))
```

```{r insertion sites}
insSite <- c(gr$insSite[sapply(gr$insSite$txs, length) == 1], 
             gr$insSite[sapply(gr$insSite$txs, length) == 2],
             gr$insSite[sapply(gr$insSite$txs, length) == 2])
insSite$tx_name <- c(unlist(gr$insSite[sapply(gr$insSite$txs, length) == 1]$txs),
                     sapply(gr$insSite[sapply(gr$insSite$txs, length) == 2]$txs, function(x){x[1]}),
                     sapply(gr$insSite[sapply(gr$insSite$txs, length) == 2]$txs, function(x){x[2]}))
insSite.bedpe <- breakpointgr2bedpe(gr$insSite) %>% 
    left_join(as_tibble(insSite)[,c("sourceId", "tx_name")], by = c("name" = "sourceId")) %>% 
    mutate(tx_name = ifelse(is.na(tx_name), "uc003yia.3", tx_name)) %>% 
    left_join(as_tibble(txs.plot)[,c("seqnames","tx_name")]) %>% 
    mutate(inschr = ifelse(as.character(seqnames)==as.character(chrom1), "chrom1", "chrom2")) %>% 
    mutate(inssite = ifelse(as.character(seqnames)==as.character(chrom1), paste0("chr",chrom2,":",start2),
                            paste0("chr",chrom1,":",start1)))
```
```{r}
library(ggrepel)
ggplot() +
    geom_segment(data=as_tibble(txs.plot), aes(x=start, xend=end, y=1.5, yend=1.5), colour="black") +
    geom_text(data=as_tibble(txs.plot), aes(x=0.9*start+0.1*end, y=3, label=paste0("chr",seqnames)))+
    geom_rect(data=as_tibble(exons.plot), mapping=aes(xmin=start, xmax = end, ymin=1, ymax = 2), colour="black") +
    # geom_point(data=g, aes(x=start, y=3, colour=factor(call))) +
    geom_curve(data=as_tibble(intronic.dels.bedpe), aes(x=start1, y=2, xend=start2, yend=2), curvature = -0.3, size=1) +
    geom_point(data=insSite.bedpe[insSite.bedpe$inschr=="chrom1",], aes(x=start1, y=1)) +
    geom_label_repel(data=insSite.bedpe[insSite.bedpe$inschr=="chrom1",], aes(x=start1, y=1, label=inssite)) +
    geom_point(data=insSite.bedpe[insSite.bedpe$inschr=="chrom2",], aes(x=start2, y=1)) +
    geom_label_repel(data=insSite.bedpe[insSite.bedpe$inschr=="chrom2",], aes(x=start2, y=1, label=inssite)) +
    # geom_point(data=insSite.bedpe, aes(x=start1, y=0, colour=chrom1)) +
    facet_wrap(~tx_name, scales = "free_x") +
    theme_bw() + scale_y_continuous(limits = c(0, 4))
    
```

